### PR TITLE
fix(Anchor): The updated onChange method is not used

### LINKED ISSUE #41925: `Anchor.onChange` is not updatable
### Reproduction link

[https://codepen.io/SkywaveTM/pen/ZEqBwKQ](https://codepen.io/SkywaveTM/pen/ZEqBwKQ)

### Steps to reproduce

1. scroll and check the counter is increasing
2. click "Change Callback"
3. scroll and check the counter is **still increasing**

### What is expected?

Since a callback is changed, the counter should decrease.

### What is actually happening?

A new callback is not reflected to an Anchor component.

| Environment | Info |
| --- | --- |
| antd | 5.4.4 |
| React | 18 |
| System | MacOS 12.5 |
| Browser | Chrome 111.0.5563.64 |

---
https://github.com/ant-design/ant-design/blob/40fb753349c4f2be314c91dbb7e6f1a960097c19/components/anchor/Anchor.tsx#L243

It seems that `onChange` should be added to the dependency list.

### PR DESCRIPTION
**Pull request by Yuiai01**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

close https://github.com/ant-design/ant-design/issues/41925

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  Fix the updated onChange method is not used           |
| 🇨🇳 Chinese | 修复没有使用更新后的 onChange 方法          |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at dd22dc4</samp>

Fixed a bug in the `Anchor` component that caused it to ignore the latest `onChange` prop from the parent component. Added a test case to verify the fix and updated the import order in `Anchor.test.tsx` and `Anchor.tsx` to follow the ESLint rule.

### <samp>🤖 Generated by Copilot at dd22dc4</samp>

*  Add `onChange` prop to dependency arrays of hooks that use it ([link](https://github.com/ant-design/ant-design/pull/41934/files?diff=unified&w=0#diff-353746edad3f07acb175d7e32c64972b87359b5a0dc4e14e4d3f1133b61e48a8L243-R243), [link](https://github.com/ant-design/ant-design/pull/41934/files?diff=unified&w=0#diff-353746edad3f07acb175d7e32c64972b87359b5a0dc4e14e4d3f1133b61e48a8L270-R270), [link](https://github.com/ant-design/ant-design/pull/41934/files?diff=unified&w=0#diff-353746edad3f07acb175d7e32c64972b87359b5a0dc4e14e4d3f1133b61e48a8L321-R321))
* Change import order and add new import for `Button` component ([link](https://github.com/ant-design/ant-design/pull/41934/files?diff=unified&w=0#diff-8403f43a04d69820d83b56e860ecd221192c3094a657c45d7c6fb3291fc48ca3L1-R7))
* Add new test case to verify dynamic `onChange` prop update ([link](https://github.com/ant-design/ant-design/pull/41934/files?diff=unified&w=0#diff-8403f43a04d69820d83b56e860ecd221192c3094a657c45d7c6fb3291fc48ca3R377-R423))

### CODE REVIEW

File: components/anchor/Anchor.tsx [hunk 1]
```diff
@@ -267,7 +267,7 @@ const AnchorContent: React.FC<InternalAnchorProps> = (props) => {
         },
       });
     },
-    [targetOffset, offsetTop],
+    [targetOffset, offsetTop, onChange],
```
Comment (afc163) [inline components/anchor/Anchor.tsx:270]: 这个 onChange 会一直为 true。

File: components/anchor/Anchor.tsx [hunk 2]
```diff
@@ -240,7 +240,7 @@ const AnchorContent: React.FC<InternalAnchorProps> = (props) => {
       bounds,
     );
     setCurrentActiveLink(currentActiveLink);
```
Comment (zombieJ) [inline components/anchor/Anchor.tsx:242]: setCurrentActiveLink 直接用 `rc-util` 的 `useEvent` 包一下就好

File: components/anchor/Anchor.tsx (hunk 1 above)
Comment (yoyo837) [reply components/anchor/Anchor.tsx:270]: 字面量函数, 每次render都是新的

Comment (Yuiai01) [reply components/anchor/Anchor.tsx:270]: 学到了

File: components/anchor/Anchor.tsx (hunk 2 above)
Comment (Yuiai01) [reply components/anchor/Anchor.tsx:242]: 好～

File: components/anchor/Anchor.tsx [hunk 3]
```diff
@@ -240,7 +240,7 @@ const AnchorContent: React.FC<InternalAnchorProps> = (props) => {
...
-  }, [dependencyListItem, targetOffset, offsetTop]);
+  }, [dependencyListItem, targetOffset, offsetTop, onChange]);
```
Comment (afc163) [inline components/anchor/Anchor.tsx:243]: 这个是一样的。

Comment (Yuiai01) [reply components/anchor/Anchor.tsx]: > 这个是一样的。

好了～ 我删掉了

**Review by MadCcc — APPROVED** [review]

**Pull request merged**